File: src/main/java/org/cubexmc/metro/service/CommandDisplayService.java
package org.cubexmc.metro.service;

import java.util.List;
import java.util.function.Function;

/**
 * Shared display helpers for command views.
 */
public class CommandDisplayService {

    public static final int DEFAULT_PAGE_SIZE = 8;

    public record Page<T>(List<T> items, int page, int totalPages, int totalItems, int pageSize) {
    }

    public record HelpPage(String header, List<String> lines, int page, int totalPages) {
    }

    public record HelpSection(String header, List<String> lines) {
    }

    public <T> Page<T> paginate(List<T> items, Integer requestedPage) {
        return paginate(items, requestedPage, DEFAULT_PAGE_SIZE);
    }

    public <T> Page<T> paginate(List<T> items, Integer requestedPage, int pageSize) {
        if (pageSize <= 0) {
            throw new IllegalArgumentException("pageSize must be greater than zero");
        }

        List<T> safeItems = items == null ? List.of() : items;
        int totalItems = safeItems.size();
        int totalPages = Math.max(1, (int) Math.ceil((double) totalItems / pageSize));
        int page = requestedPage == null ? 1 : requestedPage;
        if (page < 1) {
            page = 1;
        } else if (page > totalPages) {
            page = totalPages;
        }

        int start = (page - 1) * pageSize;
        int end = Math.min(start + pageSize, totalItems);
        List<T> pageItems = totalItems == 0 ? List.of() : List.copyOf(safeItems.subList(start, end));
        return new Page<>(pageItems, page, totalPages, totalItems, pageSize);
    }

    public HelpPage helpPage(Function<String, String> messageResolver,
                             String headerKey,
                             List<String> lineKeys,
                             Integer requestedPage) {
        Page<String> keyPage = paginate(lineKeys, requestedPage);
        List<String> lines = keyPage.items().stream()
                .map(messageResolver)
                .toList();
        String header = pageHeader(messageResolver.apply(headerKey), keyPage);
        return new HelpPage(header, lines, keyPage.page(), keyPage.totalPages());
    }

    public String pageHeader(String header, Page<?> page) {
        return header + " §e(" + page.page() + "/" + page.totalPages() + ")";
    }

    public HelpSection helpSection(Function<String, String> messageResolver,
                                   String headerKey,
                                   List<String> lineKeys) {
        List<String> lines = (lineKeys == null ? List.<String>of() : lineKeys).stream()
                .map(messageResolver)
                .toList();
        return new HelpSection(messageResolver.apply(headerKey), lines);
    }
}

File: src/main/java/org/cubexmc/metro/service/CommandDisplayService.kt
package org.cubexmc.metro.service

import java.util.function.Function

/**
 * Shared display helpers for command views.
 */
class CommandDisplayService {
    class Page<T>(
        private val items: List<T>,
        private val page: Int,
        private val totalPages: Int,
        private val totalItems: Int,
        private val pageSize: Int,
    ) {
        fun items(): List<T> = items
        fun page(): Int = page
        fun totalPages(): Int = totalPages
        fun totalItems(): Int = totalItems
        fun pageSize(): Int = pageSize
    }

    class HelpPage(
        private val header: String,
        private val lines: List<String>,
        private val page: Int,
        private val totalPages: Int,
    ) {
        fun header(): String = header
        fun lines(): List<String> = lines
        fun page(): Int = page
        fun totalPages(): Int = totalPages
    }

    class HelpSection(
        private val header: String,
        private val lines: List<String>,
    ) {
        fun header(): String = header
        fun lines(): List<String> = lines
    }

    fun <T> paginate(items: List<T>?, requestedPage: Int?): Page<T> = paginate(items, requestedPage, DEFAULT_PAGE_SIZE)

    fun <T> paginate(items: List<T>?, requestedPage: Int?, pageSize: Int): Page<T> {
        if (pageSize <= 0) {
            throw IllegalArgumentException("pageSize must be greater than zero")
        }

        val safeItems = items ?: emptyList()
        val totalItems = safeItems.size
        val totalPages = maxOf(1, kotlin.math.ceil(totalItems / pageSize.toDouble()).toInt())
        var page = requestedPage ?: 1
        if (page < 1) {
            page = 1
        } else if (page > totalPages) {
            page = totalPages
        }

        val start = (page - 1) * pageSize
        val end = minOf(start + pageSize, totalItems)
        val pageItems = if (totalItems == 0) emptyList() else java.util.List.copyOf(safeItems.subList(start, end))
        return Page(pageItems, page, totalPages, totalItems, pageSize)
    }

    fun helpPage(
        messageResolver: Function<String, String>,
        headerKey: String,
        lineKeys: List<String>,
        requestedPage: Int?,
    ): HelpPage {
        val keyPage = paginate(lineKeys, requestedPage)
        val lines = keyPage.items().map { key -> messageResolver.apply(key) }
        val header = pageHeader(messageResolver.apply(headerKey), keyPage)
        return HelpPage(header, lines, keyPage.page(), keyPage.totalPages())
    }

    fun pageHeader(header: String, page: Page<*>): String = header + " §e(" + page.page() + "/" + page.totalPages() + ")"

    fun helpSection(messageResolver: Function<String, String>, headerKey: String, lineKeys: List<String>?): HelpSection {
        val lines = (lineKeys ?: emptyList()).map { key -> messageResolver.apply(key) }
        return HelpSection(messageResolver.apply(headerKey), lines)
    }

    companion object {
        const val DEFAULT_PAGE_SIZE: Int = 8
    }
}
